Utils/azuretools-core/src/com/microsoft/azuretools/ui/embeddedbrowser/JxBrowserUtil.java

```diff
@@ -22,42 +22,303 @@
+    private static final String AZURE_ACCOUNT_NAME = "ltiantest";
+    private static final String AZURE_BLOB_URI = "http://ltiantest.blob.core.windows.net";
+    private static final String JXBROWSER_LICENSE_FILE = "license.jar";
+    private static final String JXBROWSER_COMMON_JAR = "jxbrowser-6.16.jar";
+    private static final int RETRY_BACKOFF_FACTOR = 3;
+
+    /**
+     * Async version of createBrowserViewAndLoadURL, return FutureTask<JComponent>
+     * @param url
+     * @return
+     */
+    public static FutureTask<JComponent> createBrowserViewAndLoadURLAsync(@NotNull final String url, final String targetPath) {
+        FutureTask<JComponent> createAndLoadTask = new FutureTask<> (new Callable<JComponent>() {
+            @Override
+            public JComponent call() throws JxBrowserException {
+                return createBrowserViewAndLoadURL(url, targetPath);
+            }
+        });
+
+        Thread thread = new Thread(createAndLoadTask);
+        thread.setDaemon(true);
+        thread.start();
 
-    private static final Map<String, String> versionMap = Collections.unmodifiableMap(
-        new HashMap<String, String>() {{
-            put("win32", "jxbrowser-win32-6.16");
-            put("win64", "jxbrowser-6.16");
-            put("mac", "jxbrowser-mac-6.16");
-            put("linux64", "jxbrowser-linux64-6.16");
-            put("linux32", "jxbrowser-linux64-6.16");
-        }}
-    );
-
-    public JComponent createBrowserViewAndLoadURL(String url) throws Exception {
+        return createAndLoadTask;
+    }
+
+    /**
+     * Create Swing JComponent, load the URL
+     * @param url
+     * @return JComponent (JxBrowser BrowserView) that can be added to Swing UI
+     * @throws Exception Re-throw exception like ClassNotFound, etc.
+     */
+    private static JComponent createBrowserViewAndLoadURL(@NotNull String url, String targetPath) throws JxBrowserException {
         Object browserView = null;
         try {
-            Class<?> browserClass = Class.forName("com.teamdev.jxbrowser.chromium.Browser");
+            Path jxBrowserJarPath = Paths.get(targetPath, getJxBrowserJarFileName());
+            Path licenseJarPath = Paths.get(targetPath, JXBROWSER_LICENSE_FILE);
+            Path commonJarPath = Paths.get(targetPath, JXBROWSER_COMMON_JAR);
+            File jxBrowserJarFile = jxBrowserJarPath.toFile();
+            File licenseJarFile = licenseJarPath.toFile();
+            File commonJarFile = commonJarPath.toFile();
+            if (!licenseJarFile.exists() || !licenseJarFile.isFile() || !jxBrowserJarFile.exists() || !jxBrowserJarFile.isFile() || !commonJarFile.exists() || !commonJarFile.isFile()) {
+                FileUtils.deleteQuietly(jxBrowserJarFile);
+                FileUtils.deleteQuietly(licenseJarFile);
+                FileUtils.deleteQuietly(commonJarFile);
+                downloadFilesFromAzure(new ArrayList<Path>(){{
```

Comment:
`Arrays.asList(jxBrowserJarPath, commJarPath, licenseJarPath)` might be simple.

Reply:
Fixed

---
In reply to: [156554898](https://github.com/Microsoft/azure-tools-for-java/pull/1296#discussion_r156554898) [](ancestors = 156554898)